go/install/update.go

```diff
@@ -0,0 +1,377 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package install
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/keybase/client/go/install/sources"
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol"
+	zip "github.com/keybase/client/go/tools/zip"
+
+	"github.com/blang/semver"
+)
+
+type Updater struct {
+	libkb.Contextified
+	config keybase1.UpdateConfig
+	source sources.UpdateSource
+	ticker *time.Ticker
+}
+
+func NewUpdater(g *libkb.GlobalContext, config keybase1.UpdateConfig, source sources.UpdateSource) Updater {
+	g.Log.Debug("New updater with config: %#v", config)
+	return Updater{
+		Contextified: libkb.NewContextified(g),
+		config:       config,
+		source:       source,
+	}
+}
+
+func (u Updater) Config() keybase1.UpdateConfig {
+	return u.config
+}
+
+func (u *Updater) CheckForUpdate() (update *keybase1.Update, err error) {
+	u.G().Log.Info("Checking for update, current version is %s", u.config.Version)
+
+	currentSemVersion, err := semver.Make(u.config.Version)
+	if err != nil {
+		return
+	}
+
+	update, err = u.source.FindUpdate(u.config)
+	if err != nil || update == nil {
+		return
+	}
+
+	u.G().Log.Info("Found update with version: %s", update.Version)
+	updateSemVersion, err := semver.Make(update.Version)
+	if err != nil {
+		return
+	}
+	if updateSemVersion.EQ(currentSemVersion) {
+		// Versions are the same, we are up to date
+		return
+	} else if updateSemVersion.LT(currentSemVersion) {
+		err = fmt.Errorf("Update is older version: %s < %s", updateSemVersion, currentSemVersion)
+		return
+	}
+
+	return
+}
+
+func computeEtag(path string) (string, error) {
+	var result []byte
+	file, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	hash := md5.New()
+	if _, err := io.Copy(hash, file); err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(hash.Sum(result)), nil
+}
+
+func (u *Updater) pathForFilename(filename string) string {
+	return path.Join(os.TempDir(), "KeybaseUpdates", filename)
+}
+
+func isFileScheme(url string) (string, bool) {
+	scheme := "file://"
+	if strings.HasPrefix(url, scheme) {
```

Comment:
The [Golang URL library](https://golang.org/pkg/net/url/#URL) has these features.